File: java/src/main/java/br/com/emendes/problems/NQueensII.java
package br.com.emendes.problems;

import java.util.Arrays;

/**
 * O quebra-cabeças N-Queens é um problema de por n queens em um tabuleiro de nxn,
 * em que duas rainhas não devem atacar uma a outra.<br><br>
 * <p>
 * Dado um inteiro n, retorne o número de distintas soluções para o quebra-cabeças N-Queens.<br><br>
 * <p>
 * Restrições:<br>
 * 1 <= n <= 9
 */
public class NQueensII {

  private int answer = 0;

  public int totalNQueens(int n) {
    this.answer = 0;
    char[][] board = new char[n][n];

    for (int i = 0; i < n; i++)
      Arrays.fill(board[i], '.');

    putQueens(n, 0, new boolean[n], new boolean[2 * n - 1], new boolean[2 * n - 1], board);
    return answer;
  }

  private void putQueens(int n, int i, boolean[] cols, boolean[] diag1, boolean[] diag2, char[][] board) {
    if (i == n) {
      answer += 1;
      return;
    }
    for (int j = 0; j < n; j++) {
      if (cols[j] || diag1[i + j] || diag2[j - i + n - 1])
        continue;
      board[i][j] = 'Q';
      cols[j] = diag1[i + j] = diag2[j - i + n - 1] = true;
      putQueens(n, i + 1, cols, diag1, diag2, board);
      cols[j] = diag1[i + j] = diag2[j - i + n - 1] = false;
      board[i][j] = '.';
    }
  }

}

File: javascript/problems/NQueensII.js
/**
 * O quebra-cabeça das n rainhas é o problema de colocar n rainhas em um
 * tabuleiro de xadrez nx n de forma que nenhuma das duas rainhas ataque uma à outra.
 *
 * Dado um inteiro n, retorne o número de soluções distintas para o quebra-cabeça das n rainhas.
 *
 * Restrições:
 *
 * 1 <= n <= 9
 *
 * @param {number} n
 * @return {number}
 */
const totalNQueens = function (n) {
  if (n === 1) return 1;
  if (n === 2 || n === 3) return 0;
  const diag1 = createCache(n * 2 - 1);
  const diag2 = createCache(n * 2 - 1);
  const columns = createCache(n);

  return countBoards(0, 0, n, diag1, diag2, columns, 0);
};

const countBoards = (row, queens, total, diag1, diag2, columns, counter) => {
  if (queens === total) {
    return counter + 1;
  }
  for (let j = 0; j < total; j++) {
    const diag1Index = row - j + total - 1;
    if (diag1[diag1Index] && diag2[row + j] && columns[j]) {
      diag1[diag1Index] = false;
      diag2[row + j] = false;
      columns[j] = false;
      counter = countBoards(
        row + 1,
        queens + 1,
        total,
        diag1,
        diag2,
        columns,
        counter
      );
      diag1[diag1Index] = true;
      diag2[row + j] = true;
      columns[j] = true;
    }
  }

  return counter;
};

const createCache = (length) => {
  return new Array(length).fill(true);
};
